File: solutions/java/0381.java
class RandomizedCollection {
  public boolean insert(int val) {
    if (!map.containsKey(val))
      map.put(val, new LinkedHashSet<>());
    map.get(val).add(vals.size());
    vals.add(val);
    return map.get(val).size() == 1;
  }

  public boolean remove(int val) {
    if (!map.containsKey(val) || map.get(val).isEmpty())
      return false;
    int index = map.get(val).iterator().next();
    int lastVal = vals.get(vals.size() - 1);
    vals.set(index, lastVal);
    vals.remove(vals.size() - 1);
    map.get(val).remove(index);
    map.get(lastVal).add(index);
    map.get(lastVal).remove(vals.size());
    return true;
  }

  public int getRandom() {
    int index = rand.nextInt(vals.size());
    return vals.get(index);
  }

  private List<Integer> vals = new ArrayList<>();
  private Map<Integer, LinkedHashSet<Integer>> map = new HashMap<>();
  private Random rand = new Random();
}

File: solutions/cpp/0381.cpp
class RandomizedCollection {
 public:
  bool insert(int val) {
    map[val].push_back(vals.size());
    vals.emplace_back(val, map[val].size() - 1);
    return map[val].size() == 1;
  }

  bool remove(int val) {
    if (!map.count(val) || map[val].empty()) return false;
    int index = map[val].back();
    map[vals.back().first][vals.back().second] = index;
    map[val].pop_back();
    swap(vals[index], vals.back());
    vals.pop_back();
    return true;
  }

  int getRandom() {
    int index = rand() % vals.size();
    return vals[index].first;
  }

 private:
  vector<pair<int, int>> vals;
  unordered_map<int, vector<int>> map;
};